pkg/lib/ratelimit/ratelimits.go

```diff
@@ -75,91 +75,111 @@ const (
-	VerifySIWEPerIP                  BucketName = "VerifySIWEPerIP"
-	ForgotPasswordTriggerEmailPerIP  BucketName = "ForgotPasswordTriggerEmailPerIP"
-	ForgotPasswordValidateEmailPerIP BucketName = "ForgotPasswordValidateEmailPerIP"
-	ForgotPasswordTriggerSMSPerIP    BucketName = "ForgotPasswordTriggerSMSPerIP"
-	ForgotPasswordValidateSMSPerIP   BucketName = "ForgotPasswordValidateSMSPerIP"
-
-	MessagingSMSPerIP   BucketName = "MessagingSMSPerIP"
-	MessagingEmailPerIP BucketName = "MessagingEmailPerIP"
+	VerifyPasswordPerIP            BucketName = "VerifyPasswordPerIP"
+	VerifyPasswordPerUserPerIP     BucketName = "VerifyPasswordPerUserPerIP"
+	VerifyTOTPPerIP                BucketName = "VerifyTOTPPerIP"
+	VerifyTOTPPerUserPerIP         BucketName = "VerifyTOTPPerUserPerIP"
+	VerifyPasskeyPerIP             BucketName = "VerifyPasskeyPerIP"
+	VerifyRecoveryCodePerIP        BucketName = "VerifyRecoveryCodePerIP"
+	VerifyRecoveryCodePerUserPerIP BucketName = "VerifyRecoveryCodePerUserPerIP"
+	VerifyDeviceTokenPerIP         BucketName = "VerifyDeviceTokenPerIP"
+	VerifyDeviceTokenPerUserPerIP  BucketName = "VerifyDeviceTokenPerUserPerIP"
+	VerifySIWEPerIP                BucketName = "VerifySIWEPerIP"
+
+	ForgotPasswordTriggerEmailPerIP    BucketName = "ForgotPasswordTriggerEmailPerIP"
+	ForgotPasswordTriggerSMSPerIP      BucketName = "ForgotPasswordTriggerSMSPerIP"
+	ForgotPasswordTriggerWhatsappPerIP BucketName = "ForgotPasswordTriggerWhatsappPerIP"
+
+	ForgotPasswordValidateEmailPerIP    BucketName = "ForgotPasswordValidateEmailPerIP"
+	ForgotPasswordValidateSMSPerIP      BucketName = "ForgotPasswordValidateSMSPerIP"
+	ForgotPasswordValidateWhatsappPerIP BucketName = "ForgotPasswordValidateWhatsappPerIP"
+
+	ForgotPasswordCooldownEmail    BucketName = "ForgotPasswordCooldownEmail"
+	ForgotPasswordCooldownSMS      BucketName = "ForgotPasswordCooldownSMS"
+	ForgotPasswordCooldownWhatsapp BucketName = "ForgotPasswordCooldownWhatsapp"
+
+	MessagingSMS          BucketName = "MessagingSMS"
+	MessagingSMSPerIP     BucketName = "MessagingSMSPerIP"
+	MessagingSMSPerTarget BucketName = "MessagingSMSPerTarget"
+
+	MessagingEmail          BucketName = "MessagingEmail"
+	MessagingEmailPerIP     BucketName = "MessagingEmailPerIP"
+	MessagingEmailPerTarget BucketName = "MessagingEmailPerTarget"
 
 	PresignImageUploadRequestPerUser BucketName = "PresignImageUploadRequestPerUser"
 )
 
-func (n RateLimit) resolvePerIP(cfg *config.AppConfig) *config.RateLimitConfig {
-	get := func(c *config.RateLimitConfig, fallback *config.RateLimitConfig) *config.RateLimitConfig {
-		if c != nil && c.IsEnabled() {
-			return c
+func (n RateLimit) resolvePerIP(cfg *config.AppConfig, featureCfg *config.FeatureConfig) *config.RateLimitConfig {
+	get := func(c *config.RateLimitConfig, fallback *config.RateLimitConfig, featureConfig *config.RateLimitConfig) *config.RateLimitConfig {
+		effectiveConfig := c
+
+		if !effectiveConfig.IsEnabled() && fallback != nil {
```

Comment:
The meaning seems changed. In the original code, it was checking `config.enabled == nil`, if it it true, then `fallback` is used. Here, if the rate limit was explicitly disabled, the fallback is used. This means we cannot explicitly disable a rate limit **AND** not falling back to the `fallback`.

Reply:
You are right, let me fix this.